File: apps/macos/HTTPFileDownloader.swift
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import CryptoKit
import Foundation

/// Downloads one file, reporting bytes as they arrive.
///
/// `URLSession.download(from:)` is shorter but reports nothing until it
/// finishes, and it writes to the system temp directory rather than the models
/// folder. On the self-hosted path that produced a download that looked dead:
/// the progress bar advances only when a whole file completes, and one file is
/// nearly the entire model, so for minutes nothing moved — while the disk
/// monitor, watching a models folder that was not growing, announced that the
/// connection had stalled. Both were wrong; the transfer was at full speed.
///
/// Streaming the body with `URLSession.bytes` would put the bytes where they
/// belong, but `AsyncBytes` yields one byte at a time: measured against the
/// real server it could not manage 100 MB in two minutes, where `curl` did
/// 200 MB in 23 s. So this uses a download task with a delegate, which keeps
/// URLSession's own transfer performance and adds the progress callbacks that
/// were missing.
/// Why a download that arrived intact is still not usable.
enum DownloadError: LocalizedError {
    case checksumMismatch(file: String, expected: String, actual: String)

    var errorDescription: String? {
        switch self {
        case .checksumMismatch(let file, let expected, let actual):
            // The digests are truncated: twelve characters is plenty to tell
            // two files apart by eye, and sixty-four in an alert is a wall.
            // The log has already recorded them in full.
            "\(file) downloaded but did not match the checksum your server "
            + "published (expected \(expected.prefix(12))…, got "
            + "\(actual.prefix(12))…). The file may be corrupt or still "
            + "uploading. Try again, and re-check the manifest if it persists."
        }
    }
}

final class HTTPFileDownloader: NSObject, URLSessionDownloadDelegate, @unchecked Sendable {
    /// Written to while in flight, renamed on success — the same extension
    /// `hasCompleteModel` treats as an unfinished download, so an interrupted
    /// transfer can never be mistaken for a complete model.
    static let partialExtension = "incomplete"

    private let destination: URL
    /// Lowercase hex SHA-256 the finished file must match, when the manifest
    /// supplied one. nil means "no expectation" — a manifest.txt server, or a
    /// file the manifest does not list a digest for.
    private let expectedSHA256: String?
    private let onProgress: @Sendable (Int64, Int64) -> Void

    /// Guards the continuation. The delegate can report completion twice —
    /// `didFinishDownloadingTo` then `didCompleteWithError(nil)` — and resuming
    /// a continuation twice is a crash, not a warning.
    private let lock = NSLock()
    private var continuation: CheckedContinuation<Int, Error>?
    private var session: URLSession?

    private init(destination: URL, expectedSHA256: String?,
                 onProgress: @escaping @Sendable (Int64, Int64) -> Void) {
        self.destination = destination
        self.expectedSHA256 = expectedSHA256?.lowercased()
        self.onProgress = onProgress
    }

    /// Downloads `url` to `destination`, returning the HTTP status code.
    ///
    /// A non-200 is returned rather than thrown: some files in a model are
    /// optional, and the caller is the one that knows which.
    @discardableResult
    static func download(
        from url: URL,
        to destination: URL,
        expectedSHA256: String? = nil,
        onProgress: @escaping @Sendable (Int64, Int64) -> Void
    ) async throws -> Int {
        let downloader = HTTPFileDownloader(destination: destination,
                                            expectedSHA256: expectedSHA256,
                                            onProgress: onProgress)
        return try await downloader.run(url: url)
    }

    /// SHA-256 of a file, read in chunks.
    ///
    /// Chunked because the weights are gigabytes: `Data(contentsOf:)` would
    /// map the whole 3.86 GB file to hash it. 1 MB is large enough that the
    /// read syscalls disappear against the hashing itself.
    static func sha256Hex(of url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var hasher = SHA256()
        while let chunk = try handle.read(upToCount: 1 << 20), !chunk.isEmpty {
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    private func run(url: URL) async throws -> Int {
        let queue = OperationQueue()
        queue.maxConcurrentOperationCount = 1
        let session = URLSession(configuration: .default, delegate: self, delegateQueue: queue)
        self.session = session
        defer { session.finishTasksAndInvalidate() }

        let task = session.downloadTask(with: url)
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                lock.lock()
                self.continuation = continuation
                lock.unlock()
                task.resume()
            }
        } onCancel: {
            task.cancel()
        }
    }

    private func finish(_ result: Result<Int, Error>) {
        lock.lock()
        let continuation = self.continuation
        self.continuation = nil
        lock.unlock()
        continuation?.resume(with: result)
    }

    // MARK: URLSessionDownloadDelegate

    func urlSession(
        _ session: URLSession,
        downloadTask: URLSessionDownloadTask,
        didWriteData bytesWritten: Int64,
        totalBytesWritten: Int64,
        totalBytesExpectedToWrite: Int64
    ) {
        onProgress(totalBytesWritten, totalBytesExpectedToWrite)
    }

    func urlSession(
        _ session: URLSession,
        downloadTask: URLSessionDownloadTask,
        didFinishDownloadingTo location: URL
    ) {
        let code = (downloadTask.response as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 200 else {
            finish(.success(code))
            return
        }
        // The file at `location` is deleted as soon as this method returns, so
        // it has to be moved here and now, not in the async caller.
        do {
            let fm = FileManager.default
            try fm.createDirectory(at: destination.deletingLastPathComponent(),
                                   withIntermediateDirectories: true)
            let partial = destination.appendingPathExtension(Self.partialExtension)
            if fm.fileExists(atPath: partial.path) { try fm.removeItem(at: partial) }
            try fm.moveItem(at: location, to: partial)
            // The rename below is what declares this file trustworthy —
            // `hasCompleteModel` treats anything still carrying .incomplete as
            // an interrupted download. So the digest is checked here, in
            // between: a file that fails is removed and never becomes a
            // destination path, which means a retry re-fetches it rather than
            // finding it already present and skipping it.
            if let expected = expectedSHA256 {
                let actual = try Self.sha256Hex(of: partial)
                guard actual == expected else {
                    try? fm.removeItem(at: partial)
                    finish(.failure(DownloadError.checksumMismatch(
                        file: destination.lastPathComponent,
                        expected: expected,
                        actual: actual)))
                    return
                }
            }
            if fm.fileExists(atPath: destination.path) {
                try fm.removeItem(at: destination)
            }
            try fm.moveItem(at: partial, to: destination)
            finish(.success(code))
        } catch {
            finish(.failure(error))
        }
    }

    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        didCompleteWithError error: Error?
    ) {
        // Success already resumed the continuation in didFinishDownloadingTo;
        // finish() ignores the second call.
        if let error {
            finish(.failure(error))
        }
    }
}

File: apps/dotnet/src/Core/Models/HttpFileDownloader.cs
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Net;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using Bunyi.Core.Diagnostics;

namespace Bunyi.Core.Models;

/// <summary>A downloaded file did not match the digest its server published.</summary>
public sealed class ChecksumMismatchException(string file, string expected, string actual)
    : Exception(BuildMessage(file, expected, actual))
{
    public string File { get; } = file;
    public string Expected { get; } = expected;
    public string Actual { get; } = actual;

    // Digests are truncated for the message a user sees: twelve characters is
    // plenty to tell two files apart, and sixty-four in a dialog is a wall. The
    // log records both in full (spec §10).
    private static string BuildMessage(string file, string expected, string actual) =>
        $"{file} downloaded but did not match the checksum your server published " +
        $"(expected {Short(expected)}…, got {Short(actual)}…). The file may be corrupt " +
        "or still uploading. Try again, and re-check the manifest if it persists.";

    private static string Short(string digest) =>
        digest.Length <= 12 ? digest : digest[..12];
}

/// <summary>What happened to one file.</summary>
public enum FileOutcome
{
    /// <summary>Fetched from the network.</summary>
    Downloaded,
    /// <summary>Already on disk and good — nothing transferred.</summary>
    Reused,
    /// <summary>The server does not have it, and it was not required.</summary>
    Missing,
}

/// <summary>The result of asking for one file.</summary>
public sealed record FileResult(FileOutcome Outcome, long BytesTransferred, long BytesOnDisk);

/// <summary>
/// Downloads one file, reporting bytes as they arrive and resuming a partial
/// transfer where the server allows it.
/// </summary>
public sealed class HttpFileDownloader(HttpClient http, ILogSink log)
{
    /// <summary>
    /// Extension for a transfer in flight.
    /// </summary>
    /// <remarks>
    /// The same token <see cref="ModelDownloader.Inspect"/> treats as an
    /// interrupted download, so a partial file can never be mistaken for a
    /// complete model. It is deliberately left behind on cancellation — that
    /// file is what makes resuming possible.
    /// </remarks>
    public const string PartialExtension = ".incomplete";

    private const int BufferSize = 1 << 20;   // 1 MiB: large enough that read syscalls
                                              // disappear against the hashing.

    private readonly HttpClient _http = http ?? throw new ArgumentNullException(nameof(http));
    private readonly ILogSink _log = log ?? throw new ArgumentNullException(nameof(log));

    /// <summary>
    /// Ensures <paramref name="destination"/> holds the file at
    /// <paramref name="uri"/>.
    /// </summary>
    /// <param name="expectedSha256">
    /// The digest the manifest published, or null. Where present it decides
    /// both whether an existing file may be reused and whether a fresh one is
    /// accepted — size equality is precisely the test a truncated file passes.
    /// </param>
    /// <param name="onBytes">Called with each chunk's size, as it arrives.</param>
    public async Task<FileResult> FetchAsync(
        Uri uri,
        string destination,
        string? expectedSha256,
        long? expectedSize,
        Action<long>? onBytes,
        CancellationToken ct)
    {
        ArgumentNullException.ThrowIfNull(uri);
        ArgumentException.ThrowIfNullOrWhiteSpace(destination);

        if (await CanReuseAsync(destination, expectedSha256, expectedSize, ct).ConfigureAwait(false))
        {
            var length = new FileInfo(destination).Length;
            return new FileResult(FileOutcome.Reused, 0, length);
        }

        Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
        var partial = destination + PartialExtension;

        // Resume only makes sense with a digest or a known size to check the
        // result against; without either, a resumed file could be a mixture of
        // two different server-side versions and nothing would notice.
        var resumeFrom = 0L;
        if (File.Exists(partial))
        {
            resumeFrom = new FileInfo(partial).Length;
            if (expectedSize is { } size && resumeFrom >= size) resumeFrom = 0;
        }

        using var request = new HttpRequestMessage(HttpMethod.Get, uri);
        if (resumeFrom > 0) request.Headers.Range = new RangeHeaderValue(resumeFrom, null);

        using var response = await _http
            .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct)
            .ConfigureAwait(false);

        if (response.StatusCode == HttpStatusCode.NotFound)
        {
            return new FileResult(FileOutcome.Missing, 0, 0);
        }

        response.EnsureSuccessStatusCode();

        // A 200 to a ranged request means the server ignored the range: it is
        // sending the whole file, so whatever is on disk must be discarded
        // rather than appended to.
        var appending = resumeFrom > 0 && response.StatusCode == HttpStatusCode.PartialContent;
        if (resumeFrom > 0 && !appending)
        {
            _log.Log($"{Path.GetFileName(destination)}: the server ignored the resume request, starting again.");
            resumeFrom = 0;
        }

        var hash = expectedSha256 is null ? null : IncrementalHash.CreateHash(HashAlgorithmName.SHA256);

        // Hashing what is already on disk, so the digest covers the whole file
        // and not merely the part fetched this time.
        if (appending && hash is not null)
        {
            await using var existing = File.OpenRead(partial);
            await HashStreamAsync(existing, hash, ct).ConfigureAwait(false);
        }

        var transferred = 0L;
        await using (var file = new FileStream(
            partial,
            appending ? FileMode.Append : FileMode.Create,
            FileAccess.Write,
            FileShare.None,
            BufferSize,
            useAsync: true))
        {
            await using var network = await response.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
            var buffer = new byte[BufferSize];

            while (true)
            {
                var read = await network.ReadAsync(buffer, ct).ConfigureAwait(false);
                if (read == 0) break;

                await file.WriteAsync(buffer.AsMemory(0, read), ct).ConfigureAwait(false);
                hash?.AppendData(buffer, 0, read);

                transferred += read;
                onBytes?.Invoke(read);
            }
        }

        if (hash is not null && expectedSha256 is not null)
        {
            var actual = Convert.ToHexStringLower(hash.GetHashAndReset());
            if (!string.Equals(actual, expectedSha256, StringComparison.OrdinalIgnoreCase))
            {
                // Discarded, not left for a retry to find and skip. A failed
                // file that survives is one the next run treats as present.
                TryDelete(partial);
                _log.Log(
                    $"Checksum mismatch for {Path.GetFileName(destination)}: " +
                    $"expected {expectedSha256}, got {actual}. The file was discarded.");
                throw new ChecksumMismatchException(
                    Path.GetFileName(destination), expectedSha256, actual);
            }
        }

        File.Move(partial, destination, overwrite: true);
        return new FileResult(FileOutcome.Downloaded, transferred, new FileInfo(destination).Length);
    }

    /// <summary>
    /// Whether a file already on disk can be used without fetching it again.
    /// </summary>
    /// <remarks>
    /// With a digest the digest decides; that is the whole point of publishing
    /// one, since a truncated file has exactly the size it should. Without one,
    /// size is the best test available.
    /// </remarks>
    private async Task<bool> CanReuseAsync(
        string destination, string? expectedSha256, long? expectedSize, CancellationToken ct)
    {
        if (!File.Exists(destination)) return false;

        var info = new FileInfo(destination);
        if (info.Length == 0) return false;

        if (expectedSha256 is not null)
        {
            var actual = await Sha256OfFileAsync(destination, ct).ConfigureAwait(false);
            if (string.Equals(actual, expectedSha256, StringComparison.OrdinalIgnoreCase)) return true;

            _log.Log($"Re-fetching {Path.GetFileName(destination)} — on disk but the checksum does not match.");
            return false;
        }

        if (expectedSize is { } size) return info.Length == size;

        // Nothing to check against. Keeping it is the incremental behaviour
        // §3b asks for; the completeness rule is what catches a bad file later.
        return true;
    }

    /// <summary>The SHA-256 of a file, read in chunks.</summary>
    /// <remarks>
    /// Chunked because the weights are gigabytes: reading one into memory to
    /// hash it would cost more than the download.
    /// </remarks>
    public static async Task<string> Sha256OfFileAsync(string path, CancellationToken ct)
    {
        using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        await using var stream = new FileStream(
            path, FileMode.Open, FileAccess.Read, FileShare.Read, BufferSize, useAsync: true);

        await HashStreamAsync(stream, hash, ct).ConfigureAwait(false);
        return Convert.ToHexStringLower(hash.GetHashAndReset());
    }

    private static async Task HashStreamAsync(
        Stream stream, IncrementalHash hash, CancellationToken ct)
    {
        var buffer = new byte[BufferSize];
        while (true)
        {
            var read = await stream.ReadAsync(buffer, ct).ConfigureAwait(false);
            if (read == 0) break;
            hash.AppendData(buffer, 0, read);
        }
    }

    /// <summary>
    /// Asks the server how large a file is, for a real progress total.
    /// </summary>
    /// <remarks>
    /// Two traps this has to survive. A Hugging Face <c>resolve/main</c> URL
    /// redirects to a CDN and answers with <c>Content-Length: 0</c>, publishing
    /// the real figure in <c>x-linked-size</c>; and a compressed response
    /// reports a length that will not match the bytes written to disk. In both
    /// cases the wrong answer is worse than none — it would produce a bar that
    /// never reaches its end, or reaches it early and sits there.
    /// </remarks>
    public async Task<long?> SizeOfAsync(Uri uri, CancellationToken ct)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Head, uri);
            using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode) return null;

            if (response.Headers.TryGetValues("x-linked-size", out var linked)
                && long.TryParse(linked.FirstOrDefault(), out var linkedSize)
                && linkedSize > 0)
            {
                return linkedSize;
            }

            if (response.Content.Headers.ContentEncoding.Count > 0) return null;

            var length = response.Content.Headers.ContentLength;
            return length > 0 ? length : null;
        }
        catch (HttpRequestException)
        {
            return null;   // an unknown size is not a failed download
        }
    }

    private static void TryDelete(string path)
    {
        try { if (File.Exists(path)) File.Delete(path); }
        catch (IOException) { /* a retry will overwrite it */ }
        catch (UnauthorizedAccessException) { }
    }
}
